### ISSUE TITLE
Task-private variables for `begin` statements

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#15706]
As a Chapel programmer, I would like the ability to declare task-private variables for `begin` statements.  As a motivating example:

```chapel
var A: [1..10000000] real;
begin with (var X = A[100..100000]) {
   ...
}
```

Note that potential workarounds have issues, for example:

```chapel
var A: [1..10000000] real;
var Aslice = A[100..1000000];
begin with (in Aslice) {
   ...
}
```

copies the slice twice needlessly, while also being more verbose.  Using a `ref` to avoid one of the copies:

```chapel
var A: [1..10000000] real;
ref Aslice = A[100..1000000];
begin with (in Aslice) {
   ...
}
```

has the primary downside of being more verbose, and the secondary drawback of causing a segmentation fault (I haven't tried to diagnose it).  Making the copy inside of the task:

```chapel
var A: [1..10000000] real;
begin {
   var Aslice = A[100..1000000];
}
```

can cause problems if `A` leaves scope or is reallocated before the task is scheduled whereas the task-local variable would be evaluated at task creation time.

**Comment by bradcray** [comment]
I keep running into this while working on faster / more asynchronous versions of reverse-complement for the CLBG website.

**Comment by e-kayrakli** [comment]
@bradcray in this idiom:

```chapel
const OnePerLocale = LocaleSpace dmapped Block(LocaleSpace);

var A: [OnePerLocale] [1..maxElemsPerLocale] real;

coforall loc in Locales with (ref myA = A[loc.id]) do on loc do
  writeln("locale ", here.id, " owns ", myA);
```

Would you find this or a potential alternative where the intent is on the `on` statement more intuitive? (Others or future self: we have discussed about `on` intents today, which are not supported as of today) 

While for this particular ref-based version it doesn't make much difference, think about wanting to change the intent from ref to var. If you do that with this idiom, then `myA` will be local to task that launches the `on` statement, which would make it remote within the `on` block.

ps. this is not to say that what the issue asks is not needed. I very much support the direction here.

**Comment by bradcray** [comment]
Yeah, I agree that putting it on the on-clause would definitely be more natural / less suspect.

**Comment by e-kayrakli** [comment]
Another thing I wanted to add but forgot was that if we were to do this

```chapel
coforall loc in Locales with (ref myA = A[loc.id])
```

should be allowed, while the forall version of it doesn't make sense and is not supported today

```chapel
forall loc in Locales with (ref myA = A[loc.id])
```

I don't think it is a deep issue, as a forall iteration is not a task. Or a task in a forall will likely have more than one iteration, thus different values for the induction variable. Where the story in coforall is a bit simpler.

**Comment by bradcray** [comment]
>  is not supported today

Where does it break down?

> doesn't make sense

I agree it probably doesn't do what a user might hope, but am not clear why it doesn't make sense?

Note that there's been a longstanding hope to implement a special "Locales domain map" for which `forall loc in Locales` would in fact kick off SPMD execution as a shorthand for `coforall loc in Locales do on loc` (and perhaps one that would use a spawning tree or the like at high locale counts rather than kicking them off one at a time).  If we had such support, would it change your "doesn't make sense" assessment?  The last time we tried to implement this, we couldn't land it, but IIRC, we've addressed the challenges we ran into for it since then.

**Comment by e-kayrakli** [comment]
Where it breaks down and doesn't make sense to me is that:

```chapel
forall i in 0..4*here.maxTaskPar with (var x = 2*i)
```

Now, each task will have 4 iterations of the loop, meaning `i` will not be a constant for the task. So, for any of the tasks involved, what is the value of `x`? The only way I could imagine this would work is if we make `x=2*i` at every iteration. But doesn't that go against what a task-private variable is? Rhetorically, what do you do for `forall i in myDom with (var agg = new CopyAggregator(int))`? We do not create an aggregator per iteration. 

forall over a potential distributed Locales array is interesting. And task-private variables for foralls where there's 1 iteration per task makes sense. But I am not sure if that's doable, unless I am mistaken above.

**Comment by bradcray** [comment]
Ah, so the problem is that a per-task expression is using a per-iteration value where #iterations >> #values, have I got that right?  While I agree that this isn't very useful / principled, does it break in some way, or does it just result in surprising results?    [just tried:  Currently, task-private variables are not permitted to be defined in terms of the loop induction variable.  I was guessing that it would work but be poorly behaved—like that each task got a single `x` but initialized it once using its first value of `i`, say).

> And task-private variables for foralls where there's 1 iteration per task makes sense. But I am not sure if that's doable, unless I am mistaken above.

I'm not following again.  Not doable how?

That said, I think we should split the `forall` cases here out into their own issue rather than bogging this one down (essentially: I'd prefer for nothing to get in the way of my `begin with (var i = ...)` feature request...  We could even split the cobegin and coforall cases into separate issues if they needed it...).

**Comment by e-kayrakli** [comment]
I created https://github.com/chapel-lang/chapel/issues/19618. With an answer to your question.

Back to the topic: let's have task-private variables on `begin`s!

**Comment by bradcray** [comment]
I've added a future to the OP capturing a revcomp-based example that wants this.

**Comment by bradcray** [comment]
@ronawho ran into the desire for this with an optimization on block scans this week (that said, intents on on-statements would have potentially been an even better match).
